hazelcast/compact.py

```diff
@@ -0,0 +1,70 @@
+import logging
+import typing
+
+from hazelcast.errors import HazelcastSerializationError
+from hazelcast.future import Future, ImmediateFuture
+from hazelcast.invocation import InvocationService, Invocation
+from hazelcast.protocol.codec import (
+    client_fetch_schema_codec,
+    client_send_schema_codec,
+    client_send_all_schemas_codec,
+)
+from hazelcast.serialization.compact import CompactStreamSerializer, Schema
+
+_logger = logging.getLogger(__name__)
+
+
+class CompactSchemaService:
+    def __init__(
+        self,
+        compact_serializer: CompactStreamSerializer,
+        invocation_service: InvocationService,
+    ):
+        self._compact_serializer = compact_serializer
+        self._invocation_service = invocation_service
+
+    def fetch_schema(self, schema_id: int) -> Future:
+        _logger.debug(
+            "Could not find schema with the id %s locally. It will be fetched from the cluster.",
+            schema_id,
+        )
+
+        request = client_fetch_schema_codec.encode_request(schema_id)
+        fetch_schema_invocation = Invocation(
+            request,
+            response_handler=client_fetch_schema_codec.decode_response,
+        )
+        self._invocation_service.invoke(fetch_schema_invocation)
+        return fetch_schema_invocation.future
+
+    def send_schema(self, schema: Schema, clazz: typing.Type) -> Future:
+        request = client_send_schema_codec.encode_request(schema)
+        invocation = Invocation(request)
+
+        def continuation(future):
+            future.result()
+            self._compact_serializer.register_sent_schema(schema, clazz)
+
+        self._invocation_service.invoke(invocation)
+        return invocation.future.continue_with(continuation)
+
+    def send_all_schemas(self) -> Future:
+        schemas = self._compact_serializer.get_sent_schemas()
+        if not schemas:
+            _logger.debug("There is no schema to send to the cluster.")
+            return ImmediateFuture(None)
+
+        _logger.debug("Sending the following schemas to the cluster: %s", schemas)
+
+        request = client_send_all_schemas_codec.encode_request(schemas)
+        invocation = Invocation(request, urgent=True)
+        self._invocation_service.invoke(invocation)
+        return invocation.future
+
+    def register_fetched_schema(self, schema_id: int, schema: typing.Optional[Schema]) -> None:
+        if not schema:
+            raise HazelcastSerializationError(
```

Comment:
Do you think it's a good idea to chain `SchemaNotFound` exception here so the user has more information about why the exception happened? Something like:
```python
def register_fetched_schema(self, schema_id: int, schema, err: Exception) -> None:
    if not schema:
        raise HazelcastSerializationError( ... ) from err
    # ...
```

Reply:
SchemaNotFound is meant to be a private API. Besides, it does not provide much information other than the schema id, which is already available in the exception message.